### PR TITLE
Enable TheWarehouse caching and skip double lookup

### LINKED ISSUE #15017: If the mooseapp main folder name is different from the software name, moosedocs.py do not fine the markdown files
## Bug Description
I cloned my moseapp in a folder named moseapp-new1 to perform multiple task at the same time (like running cases and developing) but if I use `moosedocs.py` to build the documentation in the moseapp-new1 folder `moose/python/mooseutils/mooseutils.py` do not find the mooseapp executable.
at `line 109` of `mooseutils.py`
```python
name = kwargs.pop('name', os.path.basename(loc))
```
initialize name as `moseapp-new1` and then at `line 124` of `mooseutils.py`:
```python
    # Search for executable with the given name
    else:
        for method in methods:
            exe_name = os.path.join(loc, name + '-' + method)
            if os.path.isfile(exe_name):
                exe = exe_name
            break
```
the class do not find the file `projects/moseapp-new1/moseapp-new1-opt` and error out.

## Steps to Reproduce
``` bash
git clone mooseapp.git mooseapp-new1

```
compile the code there and then make the documentation.

## Impact
I like to have multiple repos one for running the code that I do not touch at all for development and one for development, to keep them in the same folder I have to use two different names.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #15017

### CODE REVIEW

**Review by rwcarlsen — CHANGES_REQUESTED** [review]

File: framework/include/base/TheWarehouse.h [hunk 1]
```diff
@@ -262,13 +262,19 @@ class TheWarehouse
...
-      // if (_cache.count(_key_tup) == 0)
-      //{
-      setAttribsInner<0, KeyType<Attribs>...>(args...);
-      _cache[_key_tup] = _w->queryID(_attribs);
-      //}
-      // return _w->queryInto(_cache[_key_tup], results);
-      return _w->queryInto(_attribs, results);
+
+      size_t query_id;
+      const auto find = _cache.find(_key_tup);
```
[the remark this answers is not in the document]
Comment (loganharbour) [reply framework/include/base/TheWarehouse.h]: alright. done

**Review by rwcarlsen — APPROVED** [review]

**Comment by rwcarlsen** [comment]
What's up with the test debug petsc alt target:

```
misc/check_error.bad_second_order_test: *** ERROR ***
misc/check_error.bad_second_order_test: Unsupported order: 257
misc/check_error.bad_second_order_test: 
misc/check_error.bad_second_order_test: Stack frames: 7
misc/check_error.bad_second_order_test: 0: libMesh::print_trace(std::ostream&)
misc/check_error.bad_second_order_test: 1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
misc/check_error.bad_second_order_test: 2: void mooseError<char const*>(char const*&&)
misc/check_error.bad_second_order_test: 3: MooseApp::run()
misc/check_error.bad_second_order_test: 4: /opt/civet/build_0/moose/test/moose_test-dbg(+0x1037b) [0x5651889e537b]
misc/check_error.bad_second_order_test: 5: __libc_start_main
misc/check_error.bad_second_order_test: 6: /opt/civet/build_0/moose/test/moose_test-dbg(+0xf189) [0x5651889e4189]
misc/check_error.bad_second_order_test: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
misc/check_error.bad_second_order_test: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
misc/check_error.bad_second_order_test: :
```

**Comment by fdkong** [comment]
I am working on that crappy test

**Comment by rwcarlsen** [comment]
@fdkong - what is the status of your fix?  Can this go in yet?

**Comment by fdkong** [comment]
> @fdkong - what is the status of your fix? Can this go in yet?

@loganharbour  is working on

**Comment by loganharbour** [comment]
It's testing now. I'll merge when confident that the fix is good

**Comment by rwcarlsen** [comment]
You closed/ref'd the wrong issue with this - it should have been #15107 instead of #15017

**Comment by loganharbour** [comment]
> You closed/ref'd the wrong issue with this - it should have been #15107 instead of #15017

Whoops... thank you

**Pull request merged**